File: src/main/java/june_2025/WordLadder.java
package june_2025;

import java.util.*;

public class WordLadder {

    public static void main(String[] args) {
        String beginWord = "hit";
        String endWord = "cog";
        List<String> wordList = new ArrayList<>();
        String[] words = {"hot","dot","dog","lot","log","cog"};
        for (String w : words) {
            wordList.add(w);
        }
        WordLadder wordLadder = new WordLadder();
        int result = wordLadder.ladderLength(beginWord, endWord, wordList);
        System.out.println(result);

    }

    // O(m^2 * n) time | O(m^2 * n) space
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        if (!wordList.contains(endWord)) {
            return 0;
        }
        Map<String, List<String>> adj = new HashMap<>();
        wordList.add(beginWord);

        for (String word : wordList) {
            for (int j = 0; j < word.length(); j++) {
                String pattern = word.substring(0, j) + "*" + word.substring(j + 1);
                adj.computeIfAbsent(pattern, k -> new ArrayList<>()).add(word);
            }
        }

        Set<String> visited = new HashSet<>();
        Queue<String> queue = new LinkedList<>();
        queue.add(beginWord);
        int res = 1;

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                String word = queue.poll();
                if (word.equals(endWord)) {
                    return res;
                }
                for (int j = 0; j < word.length(); j++) {
                    String pattern = word.substring(0, j) + "*" + word.substring(j + 1);
                    List<String> neighbors = adj.get(pattern);
                    for (String nei : neighbors) {
                        if (!visited.contains(nei)) {
                            visited.add(nei);
                            queue.add(nei);
                        }
                    }
                }
            }
            res++;
        }

        return 0;
    }

}

File: python/second/word_ladder.py
class Solution(object):
    # O(m^2 * n) time | O(m^2 * n) space
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        adj = {}
        wordList.append(beginWord)

        for word in wordList:
            for i in range(len(word)):
                key = word[0:i] + "*" + word[i + 1:]
                if key not in adj:
                    adj[key] = []
                adj[key].append(word)
        
        visited = set()
        queue = []
        res = 1

        queue.append(beginWord)

        while queue:
            size = len(queue)
            for i in range(size):
                curr = queue.pop(0)
                if curr == endWord:
                    return res
                
                for j in range(len(curr)):
                    key = curr[0:j] + "*" + curr[j+1:]
                    neighbors = adj[key]
                    for nei in neighbors:
                        if nei not in visited:
                            visited.add(nei)
                            queue.append(nei)

            res += 1
        return 0

beginWord = "hit"
endWord = "cog"
wordList = ["hot","dot","dog","lot","log","cog"]

solution = Solution()
print(solution.ladderLength(beginWord, endWord, wordList))
